runtime/browser/ui/desktop/download_views.cc

```diff
@@ -0,0 +1,271 @@
+// Copyright (c) 2015 Intel Corporation. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "xwalk/runtime/browser/ui/desktop/download_views.h"
+
+#include <algorithm>
+
+#include "base/strings/utf_string_conversions.h"
+#include "grit/xwalk_resources.h"
+#include "third_party/skia/include/core/SkPaint.h"
+#include "ui/views/layout/grid_layout.h"
+#include "ui/base/resource/resource_bundle.h"
+#include "ui/gfx/canvas.h"
+#include "url/gurl.h"
+#include "xwalk/runtime/browser/ui/native_app_window_desktop.h"
+
+namespace xwalk {
+
+namespace {
+
+const int kDownloadImageHorizontalPadding = 4;
+const int kDownloadImageVerticalPadding = 4;
+const int kCloseButtonRightPadding = 4;
+const int kCloseButtonLeftPadding = 10;
+const int kDownloadItemPadding = 3;
+
+const int kMaxItemTextWidth = 160;
+const int kItemTextHorizontalPadding = 3;
+const int kItemTextVerticalPadding = 4;
+
+const SkColor kDownloadInProgressColor = SkColorSetARGB(128, 86, 167, 247);
+const SkColor kDownloadCompleteColor = SkColorSetARGB(128, 165, 214, 167);
+const SkColor kDownloadCancelledColor = SkColorSetARGB(128, 239, 154, 154);
+
+}  // namespace
+
+DownloadBarView::DownloadBarView(NativeAppWindowDesktop* app_view)
+    : parent_(app_view),
+      download_image_(nullptr),
+      ellipsis_label_(nullptr),
+      close_button_(nullptr) {
+  ui::ResourceBundle& rb = ui::ResourceBundle::GetSharedInstance();
+
+  download_image_ = new views::ImageView();
+  download_image_->SetImage(rb.GetImageSkiaNamed(IDR_DOWNLOAD));
+  AddChildView(download_image_);
+
+  ellipsis_label_ = new views::Label(base::ASCIIToUTF16("..."));
+  AddChildView(ellipsis_label_);
+
+  close_button_ = new views::ImageButton(this);
+  close_button_->SetImage(
+      views::CustomButton::STATE_NORMAL, rb.GetImageSkiaNamed(IDR_CLOSE));
+  close_button_->SetImage(
+      views::CustomButton::STATE_HOVERED, rb.GetImageSkiaNamed(IDR_CLOSE_H));
+  close_button_->SetImage(
+      views::CustomButton::STATE_PRESSED, rb.GetImageSkiaNamed(IDR_CLOSE_P));
+  AddChildView(close_button_);
+}
+
+DownloadBarView::~DownloadBarView() {
+}
+
+void DownloadBarView::Layout() {
+  views::View::Layout();
+
+  int x, y, width, height;
+  gfx::Size image_size = download_image_->GetPreferredSize();
+  width = image_size.width() + 2 * kDownloadImageHorizontalPadding;
+  height = image_size.height() + 2 * kDownloadImageVerticalPadding;
+  download_image_->SetBounds(kDownloadImageHorizontalPadding,
+      kDownloadImageVerticalPadding,
+      image_size.width(), image_size.height());
+
+  gfx::Size btn_size = close_button_->GetPreferredSize();
+  x = this->width() - btn_size.width() - kCloseButtonRightPadding;
+  y = (height - btn_size.height()) / 2;
```

Comment:
No. images used in download_image_ and close_button are fixed size and the former height is always > the latter.